### ISSUE TITLE
{"op":"add"} has no value if the value is undefined.

### ISSUE THREAD

**Issue by wolfgang42** [Starcounter-Jack/JSON-Patch#60]
RFC 6902, section 4.1 ("add") says: "The operation object MUST contain a "value" member whose content specifies the value to be added."

jsonpatch.compare() can, however, produce an add operation with no value when the value is undefined.

Test case:
```javascript
original = {foo: 1};
changed = {foo: 1, bar: undefined};
patch = jsonpatch.compare(original, changed);
jsonpatch.apply(original, patch);
```
Result: `Error: 'value' MUST be defined`

**Comment by warpech** [comment]
Please note that `undefined` is not a valid value in JSON (ECMA-404).

Theoretically we could ditch checking via `undefined` for `hasOwnProperty`, but that would have performance implications: http://jsperf.com/hasownproperty-vs-undefined so I prefer to stick to the standard

**Comment by warpech** [comment]
I think that actually the right solution is that `jsonpatch.compare` should create:

```js
["op": "remove", "path": "/bar"]
```

WDYT?

**Comment by warpech** [comment]
Related: #32

**Comment by sonnyp** [comment]
@warpech interesting.

I think we should throw an error for ```undefined```. Hard to do now as it would break lot of code though.

**Comment by sonnyp** [comment]
See https://github.com/Starcounter-Jack/Fast-JSON-Patch/issues/32#issuecomment-69583594

**Comment by wolfgang42** [comment]
Indeed, this is a duplicate of #36; I missed this because I was looking specifically for `{"op":
"add"}`. Personally, I think that it should not be possible for fast-json-patch to emit an invalid json-patch; in this case, it ought to raise an error informing the programmer that `undefined` is invalid JSON instead.

**Comment by sonnyp** [comment]
@wolfgang42 +1

**Comment by tomalec** [comment]
I've just added few test cases for it, so it shoul no longer result in error. Behavior in case of `undefined` is described at [README.md](https://github.com/Starcounter-Jack/JSON-Patch/blob/master/README.md#undefineds-json-to-js-extension), and at #32

**Comment by miyconst** [comment]
The test cases added by @tomalec does not work without native `Object.observe` support.
As stated in the issue https://github.com/Starcounter-Jack/JSON-Patch/issues/82, the `Object.observe` is no longer supported, so we can't rely on it.

Current implementation of dirty checking creates mirror ob object with the following code:

```js
mirror.value = JSON.parse(JSON.stringify(obj));
```

As mentioned above, `undefined` is not a valid JSON value, and `JSON.stringify` ignores such properties.

This leads to:

- Replacing value of an `undefined` property generates `add` patch instead of `replace`.
- Removing property with `undefined` value does not generate any patches.

**Comment by warpech** [comment]
Fixed in [1.0.0](https://github.com/Starcounter-Jack/JSON-Patch/releases/tag/1.0.0)
